### PR TITLE
Update PL/PLR visualization to use new PF pipeline package

### PR DESCRIPTION
**Pull request by karthikjeeyar**
Fixes:
https://issues.redhat.com/browse/ODC-6690
https://issues.redhat.com/browse/OCPBUGSM-44997

**Analysis / Root cause**: 

Update ODC pipeline to new PF pipeline package.

**Solution Description**: 

This PR brings in the latest PF pipeline topology, which includes
1. Zoom in/out to changes the node shapes/details
2. Standard topology icons to zoom in/out, fit and reset button controls.

**Screen shots / Gifs for design review**: 

Pipeline default view:
---

image

PipelineRun default view:
---

image

Zoomed out view:
---
image

Hovering on a node and opening tooltip:
---
image

Gif:
---
pipelinerun

complex-pipeline

**Unit test coverage report**: 

```
utils.spec.ts:
extractDepsFromContextVariables: 
    ✓ should return emtpy array for invalid values (1ms)
    ✓ should return empty array if the context variable string does not contain results
    ✓ should return a task name if the context variable string contains results
    ✓ should return a list of task names if the context variable string contains multiple results
  taskWhenStatus:
    ✓ should return undefined if the task does not have when expression
    ✓ should return a matching when status (1ms)
  getGraphDataModel
    ✓ should return null for invalid values (2ms)
    ✓ should return graph, nodes and edges for valid pipeline (5ms)
    ✓ should return graph, nodes and edges for valid pipeline (1ms)
    ✓ should include the finally group and nodes for the pipeline with finally task (1ms)
```    

```
 dag
    ✓ should print the tasks in topological sort order 
    addVertex:
      ✓ should add the vertex to the graph (2ms)
      ✓ should skip the duplicate vertex added to the graph (1ms)
    AddEdge:
      ✓ should add the vertex if it not available in the graph
      ✓ should add the edges to the graph
      ✓ should skip the duplicate edges added to the graph
    AddEdges
      ✓ should form the dependancies when addEdges is called with multiple before and after tasks (5ms)
      ✓ should throw an error if there is any cycle detected (1ms)
```
      
**Test setup:**

1. Add a resource with pipeline from Git import flow.
2. Visit Pipeline and pipelineRun Details view.

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc: @serenamarie125 @siamaksade @jeff-phillips-18

### CODE REVIEW

**Comment by karthikjeeyar** [comment]
cc: @invincibleJai @vikram-raj

**Comment by vikram-raj** [comment]
Skipped tasks icon changed

Before:
image

Now:
image

Is it expected?

**Comment by karthikjeeyar** [comment]
> Skipped tasks icon changed
> Is it expected?

@vikram-raj Good catch, it is not expected. I have fixed it

image

**Comment by karthikjeeyar** [comment]
> I'm a little bit concerned about the accessibility of the thin connecting lines and node outlines in the zoomed out view. Would love to hear what everyone else thinks

@MariaLeonova Fair point. One more point is while the task is in progress, there is no indication which makes it difficult to interpret which is nodes are running and which are not. We should bring this up with PF team and PM to see how to improve this.

**Comment by karthikjeeyar** [comment]
> Would it be helpful to know what tasks are Completed, In-progress and Failed? Or all of the above with skipped and cancelled? Trade-off is it can bury the Alerts if it gets too busy so I welcome your thoughts on this.

Completed tasks are indicated with Green outline on the node and failed/cancelled has icons that grab's users attention. 
Maybe we could have blue outline representing In-progress nodes so that users understand the flow at glance and at the same time Alerts wont get buried with many icons?

 I can bring this up in tmp-odc-pipelines channel to get feedback from PMs and other stakeholders.

**Review by vikram-raj — APPROVED** [review]
Thanks @karthikjeeyar. Verified the changes and it works as expected.

**Review by vikram-raj — COMMENTED** [review]
/lgtm

**Comment by jerolimov** [comment]
Hey @karthikjeeyar /cc @jeff-phillips-18 @serenamarie125 

Maybe you have discussed this in slack somewhere, but why is the default zoom level low for a pipeline with just three tasks?

This is how it looks like for a default pipeline that we add currently in the Import from Git flow:

https://user-images.githubusercontent.com/139310/191311293-5de47ea0-f479-4473-8962-29b2b1a1a980.mp4

Some questions or improvement ideas:

1. Shouldn't the default zoom level and view position be the same as "Reset View"? I need two clicks on "Zoom out" after "Reset view" to have the same zoom level.
2. Could we make this depending on the number of tasks?
3. What is the difference between "Fit to Screen" and "Reset view"? Both position the view to the center and set the zoom level to "normal". I would expect that "Fit to Screen" uses a smart zoom level so that everything "fits into my screen", but this is not the case on the pipeline details page:

https://user-images.githubusercontent.com/139310/191313531-43f60a71-bbb4-4a86-9742-e4c3450a631c.mp4

/hold

**Review by jerolimov — CHANGES_REQUESTED** [review]
see comment above

**Comment by karthikjeeyar** [comment]
> Some questions or improvement ideas:
> Shouldn't the default zoom level and view position be the same as "Reset View"? I need two clicks on "Zoom out" after "Reset view" to have the same zoom level. Could we make this depending on the number of tasks?

@jerolimov are you suggesting, default zoom level should be set to low only when there are more number of tasks (columns) in the graph otherwise set the zoom level to high? 


>What is the difference between "Fit to Screen" and "Reset view"? Both position the view to the center and set the zoom level to "normal". I would expect that "Fit to Screen" uses a smart zoom level so that everything "fits into my screen", but this is not the case on the pipeline details page.

Fixed it.

>Fit to screen also ignores that there are some toolbar icons on the canvas. It would be nice if the high of the toolbar icons would be "ignored" for the calculation.

This needs to be handled via PF. we can log an issue in PF to fix this.

**Comment by vikram-raj** [comment]
I noticed that we can open multiple tooltips at the same time. See the screenshot below.
image

**Comment by karthikjeeyar** [comment]
@vikram-raj These tooltip changes are coming from PF. Older topology version was not opening multiple tooltips, we need to log a bug in PF to fix it.

**Comment by karthikjeeyar** [comment]
@serenamarie125 @jerolimov I have addressed the default zoom level (reset view), added a max height of 300px to the graph and handled window resize issue as well.

Complex Pipeline example:
---
**Before:**

Screenshot 2022-09-22 at 4 59 58 PM


**After (With Fixed minimum height + new PF topology changes) default view:**
---
Screenshot 2022-09-22 at 4 59 49 PM

 **Fit to screen** mode:
 ---
image

**Comment by karthikjeeyar** [comment]
Rebased PR

**Comment by serenamarie125** [comment]
@karthikjeeyar thanks for all the detailed screenshots for the PL - they look great.  Could you add something similar for the PLR?  Sorry to be so picky :)

**Comment by karthikjeeyar** [comment]
@serenamarie125 I have updated the [comment](https://github.com/openshift/console/pull/12046#issuecomment-1254901676) with PLR visualization screenshots.

**Comment by vikram-raj** [comment]
Thanks @karthikjeeyar. Tested the PR locally and all functionality works as expected.

**Comment by karthikjeeyar** [comment]
> Can we keep the Fit to screen as a default view?

Yes, this is Ideal way to see all the tasks at a glance. @jerolimov was suggesting the same approach for default view. Default fit to screen has to be handled via PF so that we wont be doing any custom handling of this behaviour. @jeff-phillips-18 Wdyt?

**Review by vikram-raj — COMMENTED** [review]
/lgtm

**Comment by vikram-raj** [comment]
@sanketpathak / @hemantsaini-7 Please verify this PR.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx:48]: Nit:

`stepStatusList`  is always defined, and if not this will crash without optional changing, so we don't the optional check and fallback to zero here:

```suggestion
  const succeededStepsCount = stepStatusList.filter(({ status }) => status === ComputedStatus.Succeeded).length;
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx:29]: Missing the type here, but it looks like there is no type for the data that has a `task`, `taskIconClass`, `taskIcon`, etc. So we can not add it now, but we should create and request such types in the future.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/types.ts
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/types.ts:19]: I like types :smirk: 

```suggestion
  whenStatus?: WhenStatus;
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts:417]: `taskWhenStatus` returns undefined if `!task.when`, so we can do this here as well and switch the return type from string to `WhenStatus`?

```suggestion
export const getWhenStatus = (status: RunStatus): WhenStatus => {
  switch (status) {
    case RunStatus.Succeeded:
    case RunStatus.Failed:
      return WhenStatus.Met;
    case RunStatus.Skipped:
    case RunStatus['In Progress']:
    case RunStatus.Idle:
      return WhenStatus.Unmet;
    default:
      return undefined;
  }
};
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts:424]: Should we use get here as well, similar to most other getters in this file?

```suggestion
export const `getTaskWhenStatus = (task: PipelineTaskWithStatus): WhenStatus => {
  if (!task.when) {
    return undefined;
  }
  return getWhenStatus(task.status?.reason);
};
```

**Review by jerolimov — COMMENTED** [review]
Small change requests, overall it looks really good. As discussed, I will ask some more questions about my UI/UX findings in slack.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx:29]: Yes we can not add it now, the base ElementModel's data is mapped to any. so we don't have any type to add here at the moment.

**Review by jerolimov — APPROVED** [review]
Thanks for applying changes the latest comments. :+1: 

/unhold
/lgtm

**Comment by jerolimov** [comment]
Sorry, let us discuss the UX changes first before we merge this. Setting this on hold until Friday.

/hold

**Comment by karthikjeeyar** [comment]
Based on the discussion made few changes to the PR.

1. Showing the existing Tooltip in both Pipeline/PipelineRun nodes ( Will be modified once PF implements the popover on hover feature).
2. Setting the graph height to 45 percent of the window size to provide some scroll area below the graph.

Issues needs to be fixed via PF:
1. Status colour on Running/Pending icons.
2. Configuration options to hide/show the icons in zoomed node.
3. Ally issue while navigating via keyboard - highlight (black border) around the current node is missing.

**45 percent Height:**

https://user-images.githubusercontent.com/9964343/193296035-82fa07c6-9ebe-421b-a9d1-0902c6664f50.mov

**Pipeline task with Tooltip:**

https://user-images.githubusercontent.com/9964343/192962183-fe6e9470-8fc6-44a5-bfc1-77f451c88332.mov

**PipelineRun task with Tooltip:**

https://user-images.githubusercontent.com/9964343/192962359-08db7a8d-0ba1-4f5f-aed6-d87c8ff0b7f0.mov

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx]: fixed it.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx:71]: Can computedTask be undefined? Then I would recommend:

```
  const isSkipped = !!(computedTask && data?.pipelineRun?.status?.skippedTasks?.some(
    (t) => t.name === data?.task.name,
    (t) => t.name === computedTask.name,
  ));
```

Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/PipelineTaskNode.tsx:71]: Its is the main resource for this task node, it will not be undefined but I can add the check.

**Review by serenamarie125 — APPROVED** [review]
This looks great, thanks for collaborating on this!

**Pull request merged**
